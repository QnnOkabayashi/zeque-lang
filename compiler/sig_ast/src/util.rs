use std::hash::Hash;
use std::iter::{Enumerate, Rev};
use std::ops::{Deref, Index, IndexMut};
use std::{cmp, fmt, hash, marker::PhantomData, slice::Iter};

pub type StringInterner = string_interner::StringInterner<string_interner::backend::BufferBackend>;

#[derive(Copy, Clone, Debug)]
pub struct Span<T>(pub T, pub Range);

impl<T> Span<T> {
    pub fn new(start: usize, end: usize, inner: T) -> Self {
        Span(
            inner,
            Range {
                start: start as u32,
                end: end as u32,
            },
        )
    }

    pub fn range(&self) -> Range {
        self.1
    }

    pub fn into_inner(self) -> T {
        self.0
    }
}

/// PartialEq ignores range
impl<T: PartialEq> PartialEq for Span<T> {
    fn eq(&self, other: &Self) -> bool {
        self.0 == other.0
    }
}

impl<T: Eq> Eq for Span<T> {}

/// Hash ignores range
impl<T: Hash> Hash for Span<T> {
    fn hash<H: hash::Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}

#[derive(Copy, Clone, Debug)]
pub struct Range {
    pub start: u32,
    pub end: u32,
}

impl Range {
    pub fn to(self, other: Self) -> Self {
        debug_assert!(self.end <= other.start, "unexpected use of Range::to");
        Range {
            start: self.start,
            end: other.end,
        }
    }
}

// invariant: values and ranges are the same length
#[derive(Clone, Debug)]
pub struct RangeTable<T> {
    values: Vec<T>,
    ranges: Vec<Range>,
}

impl<T> RangeTable<T> {
    pub fn new() -> Self {
        RangeTable::default()
    }

    pub fn ranges(&self) -> &[Range] {
        &self.ranges
    }

    pub fn span(&self, index: Ix<T>) -> Span<&T> {
        Span(&self.values[index], self.ranges[index.index])
    }

    pub fn push(&mut self, Span(value, range): Span<T>) -> Ix<T> {
        self.ranges.push(range);
        Ix::push(&mut self.values, value)
    }
}

impl<T> Default for RangeTable<T> {
    fn default() -> Self {
        RangeTable {
            values: Vec::new(),
            ranges: Vec::new(),
        }
    }
}

impl<T> Deref for RangeTable<T> {
    type Target = [T];

    fn deref(&self) -> &Self::Target {
        &self.values
    }
}

pub struct Scope<'scope, T> {
    original_len: usize,
    stack: &'scope mut Vec<T>,
}

impl<T> Drop for Scope<'_, T> {
    fn drop(&mut self) {
        self.stack.truncate(self.original_len);
    }
}

impl<'scope, T> Scope<'scope, T> {
    pub fn new(env: &'scope mut Vec<T>) -> Self {
        Scope {
            original_len: env.len(),
            stack: env,
        }
    }

    pub fn enter_scope(&mut self) -> Scope<'_, T> {
        Scope {
            original_len: self.stack.len(),
            stack: self.stack,
        }
    }

    pub fn push(&mut self, value: T) {
        self.stack.push(value);
    }

    /// Includes the enumerate part in case people want to use the index...
    /// This is a _stack_. We search _backwards_. So if you want the position, you need to
    /// make sure the index from enumerate reflects that, which it does if .rev() comes
    /// _after_ .enumerate().
    pub fn iter(&self) -> Rev<Enumerate<Iter<'_, T>>> {
        self.stack.iter().enumerate().rev()
    }

    pub fn get(&self, index: usize) -> &T {
        &self.stack[index]
    }

    pub fn get_mut(&mut self, index: usize) -> &mut T {
        &mut self.stack[index]
    }

    pub fn in_scope(&self) -> &[T] {
        &self.stack[self.original_len..]
    }
}

impl<T> Extend<T> for Scope<'_, T> {
    fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
        self.stack.extend(iter)
    }
}

/// An index type.
pub struct Ix<I: ?Sized> {
    pub index: usize,
    _marker: PhantomData<I>,
}

impl<I> Ix<I> {
    pub fn new(index: usize) -> Self {
        Ix {
            index,
            _marker: PhantomData,
        }
    }

    pub fn map<U>(self) -> Ix<U> {
        Ix::new(self.index)
    }

    /// Push a value and get an index to it.
    pub fn push<T>(vec: &mut Vec<T>, value: T) -> Self
    where
        I: Indexes<T>,
    {
        let this = Ix::new(vec.len());
        vec.push(value);
        this
    }

    /// Iterate over the indices of a slice.
    pub fn iter(slice: &[I]) -> impl Iterator<Item = Ix<I>> {
        (0..slice.len()).map(Ix::new)
    }

    pub fn enumerate(iter: &[I]) -> impl Iterator<Item = (Self, &I)> {
        iter.iter()
            .enumerate()
            .map(|(index, value)| (Ix::new(index), value))
    }
}

impl<T> Copy for Ix<T> {}

impl<T> Clone for Ix<T> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<T> fmt::Debug for Ix<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Ix<{}>({})", std::any::type_name::<T>(), self.index)
    }
}

impl<T> fmt::Display for Ix<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

impl<T> cmp::PartialEq for Ix<T> {
    fn eq(&self, other: &Self) -> bool {
        self.index == other.index
    }
}

impl<T> cmp::Eq for Ix<T> {}

impl<T> hash::Hash for Ix<T> {
    fn hash<H: hash::Hasher>(&self, state: &mut H) {
        self.index.hash(state);
    }
}

/// If Foo implements Indexes<Bar>, then an Ix<Foo> may be used
/// to index into a &[Bar].
/// All types are allowed to index into slices containing themselves.
pub trait Indexes<T> {}

impl<T> Indexes<T> for T {}

// Can index into a slice of options,
// e.g. Ix<Function> can index into &[Option<Function>]
impl<T> Indexes<Option<T>> for T {}

impl<T, I: Indexes<T>> Index<Ix<I>> for [T] {
    type Output = T;

    fn index(&self, index: Ix<I>) -> &Self::Output {
        &self[index.index]
    }
}

impl<T, I: Indexes<T>> IndexMut<Ix<I>> for [T] {
    fn index_mut(&mut self, index: Ix<I>) -> &mut Self::Output {
        &mut self[index.index]
    }
}

impl<T, I: Indexes<T>> Index<Ix<I>> for Vec<T> {
    type Output = T;

    fn index(&self, index: Ix<I>) -> &Self::Output {
        self.as_slice().index(index)
    }
}

impl<T, I: Indexes<T>> IndexMut<Ix<I>> for Vec<T> {
    fn index_mut(&mut self, index: Ix<I>) -> &mut Self::Output {
        self.as_mut_slice().index_mut(index)
    }
}
