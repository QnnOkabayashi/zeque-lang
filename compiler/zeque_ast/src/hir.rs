pub use crate::ast::BinOp;
use index_vec::IndexVec;
use smol_str::{SmolStr, ToSmolStr};
use std::{fmt, str::FromStr};

// pub mod printer;

index_vec::define_index_type! {
    #[derive(Default)]
    pub struct ExprIdx = u32;
}

index_vec::define_index_type! {
    pub struct LetIdx = u32;
}

index_vec::define_index_type! {
    pub struct StructIdx = u32;
}

index_vec::define_index_type! {
    pub struct ParamIdx = u32;
}

index_vec::define_index_type! {
    pub struct FnIdx = u32;
}

index_vec::define_index_type! {
    pub struct ParentRefIdx = u32;
}

index_vec::define_index_type! {
    #[derive(Default)]
    pub struct LevelIdx = u32;
}

index_vec::define_index_type! {
    pub struct FileIdx = u32;
}

#[derive(Debug)]
pub struct Hir {
    pub structs: IndexVec<StructIdx, Struct>,
    pub files: IndexVec<FileIdx, File>,
    pub main: FileIdx,
}

#[derive(Clone, Debug)]
pub struct File {
    pub struct_idx: StructIdx,
    pub ctx: Ctx,
}

#[derive(Clone, Debug, Default)]
pub struct Struct {
    pub fns: IndexVec<FnIdx, FnDecl>,
    pub fields: Vec<FieldDecl>,
}

#[derive(Clone, Debug, Default)]
pub struct FnDecl {
    pub is_pub: bool,
    pub name: SmolStr,
    pub params: Vec<ParamIdx>,
    pub return_ty: ExprIdx,
    pub body: Block,
    pub ctx: Ctx,
}

#[derive(Clone, Debug, Default)]
pub struct Ctx {
    // Params live in Ctx so NameRefs can refer to them easily
    pub params: IndexVec<ParamIdx, Param>,
    // Exprs live in Ctx so they can all be stored contiguously
    pub exprs: IndexVec<ExprIdx, Expr>,
    pub lets: IndexVec<LetIdx, Let>,
    /// Names in this context can look up where to find
    /// what they're looking for in their parent context.
    pub parent_captures: IndexVec<ParentRefIdx, ParentRef>,
}

#[derive(Clone, Debug)]
pub struct Param {
    pub is_comptime: bool,
    pub name: SmolStr,
    pub ty: ExprIdx,
}

#[derive(Clone, Debug)]
pub struct Let {
    pub name: SmolStr,
    pub ty: Option<ExprIdx>,
    pub expr: ExprIdx,
}

#[derive(Clone, Debug, Default)]
pub struct Block {
    pub stmts: Vec<Stmt>,
    pub returns: ExprIdx,
}

#[derive(Copy, Clone, Debug)]
pub enum Stmt {
    Let(LetIdx),
}

#[derive(Clone, Debug)]
pub enum Expr {
    Int(i32),
    Bool(bool),
    BinOp {
        op: BinOp,
        lhs: ExprIdx,
        rhs: ExprIdx,
    },
    IfThenElse {
        cond: ExprIdx,
        then: ExprIdx,
        else_: ExprIdx,
    },
    Name(Name),
    BuiltinType(BuiltinType),
    SelfType,
    Block(Block),
    Call {
        callee: Callee,
        args: Vec<ExprIdx>,
    },
    Comptime(ExprIdx),
    Struct(StructIdx),
    Constructor {
        ty: Option<ExprIdx>,
        fields: Vec<ConstructorField>,
    },
    Field {
        expr: ExprIdx,
        field_name: SmolStr,
    },
    Error,
}

#[derive(Clone, Debug)]
pub enum Callee {
    Expr(ExprIdx),
    Builtin(BuiltinFn),
}

#[derive(Copy, Clone, Debug)]
pub enum Name {
    Local(Local),
    /// Reference to a name declaration.
    /// go look in your own context, and it will tell you where
    /// in the parent to look
    ParentRef(ParentRefIdx),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Local {
    Let(LetIdx),
    Param(ParamIdx),
    Fn(StructIdx, FnIdx),
}

#[derive(Copy, Clone, Debug)]
pub enum ParentRef {
    /// Look in my parent's layer
    Local(Local),
    /// Something my parent captured, keep going...
    Capture(ParentRefIdx),
}

#[derive(Clone, Debug)]
pub struct ConstructorField {
    pub name: SmolStr,
    pub value: ExprIdx,
}

#[derive(Clone, Debug)]
pub struct FieldDecl {
    pub name: SmolStr,
    pub ty: ExprIdx,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum BuiltinType {
    I32,
    Bool,
    Type,
}

#[derive(Clone, Debug)]
pub enum BuiltinFn {
    InComptime,
    Trap,
    Clz,
    Ctz,
    Unknown(SmolStr),
}

impl FromStr for BuiltinFn {
    type Err = std::convert::Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "in_comptime" => Ok(BuiltinFn::InComptime),
            "trap" => Ok(BuiltinFn::Trap),
            "clz" => Ok(BuiltinFn::Clz),
            "ctz" => Ok(BuiltinFn::Ctz),
            _ => Ok(BuiltinFn::Unknown(s.to_smolstr())),
        }
    }
}

impl BuiltinType {
    pub fn as_str(&self) -> &'static str {
        match self {
            BuiltinType::I32 => "i32",
            BuiltinType::Bool => "bool",
            BuiltinType::Type => "type",
        }
    }
}

impl FromStr for BuiltinType {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "i32" => Ok(BuiltinType::I32),
            "bool" => Ok(BuiltinType::Bool),
            "type" => Ok(BuiltinType::Type),
            _ => Err(()),
        }
    }
}

impl fmt::Display for BuiltinType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self.as_str(), f)
    }
}
