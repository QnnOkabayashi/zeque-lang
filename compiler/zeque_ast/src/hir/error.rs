use crate::{ast, util::Range};
use index_vec::IndexVec;
use miette::{Diagnostic, LabeledSpan};
use smol_str::SmolStr;
use std::iter;
use thiserror::Error;

index_vec::define_index_type! {
    pub struct ErrorIdx = u32;
}

pub type ErrorVec = IndexVec<ErrorIdx, NameResolutionError>;

#[derive(Clone, Debug, Error)]
#[error("name '{name}' not found")]
pub struct NameResolutionError {
    name: SmolStr,
    range: Range,
}

impl NameResolutionError {
    pub fn new(name: ast::Name) -> Self {
        NameResolutionError {
            name: name.text,
            range: name.range,
        }
    }
}

impl Diagnostic for NameResolutionError {
    fn help(&self) -> Option<Box<dyn std::fmt::Display + '_>> {
        Some(Box::new("use a name that's in scope"))
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
        Some(Box::new(iter::once(LabeledSpan::at(
            self.range,
            format!("name '{}' is not valid in this scope", self.name),
        ))))
    }
}
