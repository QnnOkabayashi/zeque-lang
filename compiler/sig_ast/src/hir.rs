//! The high-level intermediate representation.
//!
//! This IR represents a name-resolved version of a Sig program.
//! This means that we now know what every identifier refers to.
//!
//! It also flattens the tree structure into vectors stored in the [`FunctionContext`].
//! For displaying the HIR as a tree structure, see the [`printer`] module.
//!
//! All span information is stored in FunctionContext (for now).

pub use crate::ast::BinOp;
use crate::util::{Ix, RangeTable, Span};
use std::{fmt, str::FromStr};
use string_interner::DefaultSymbol;

pub mod printer;

#[derive(Clone, Debug)]
pub enum Item {
    Fn(Function),
}

#[derive(Clone, Debug)]
pub struct Function {
    pub name: Span<DefaultSymbol>,
    pub return_type: Ix<Expr>,
    pub body: Ix<Block>,
    pub context: FunctionContext,
}

#[derive(Clone, Debug, Default)]
pub struct FunctionContext {
    pub params: RangeTable<Param>,
    pub exprs: RangeTable<Expr>,
    pub lets: Vec<Let>,
    pub blocks: Vec<Block>,
    pub structs: Vec<Struct>,
}

#[derive(Clone, Debug)]
pub struct Param {
    pub is_comptime: bool,
    pub name: Span<DefaultSymbol>,
    pub ty: Ix<Expr>,
}

#[derive(Clone, Debug)]
pub struct Let {
    pub name: Span<DefaultSymbol>,
    pub ty: Option<Ix<Expr>>,
    pub expr: Ix<Expr>,
}

#[derive(Clone, Debug)]
pub struct Block {
    pub stmts: Vec<Stmt>,
    pub returns: Ix<Expr>,
}

#[derive(Copy, Clone, Debug)]
pub enum Stmt {
    Let(Ix<Let>),
}

#[derive(Clone, Debug)]
pub enum Expr {
    Int(i32),
    Bool(bool),
    BinOp(BinOp, Ix<Self>, Ix<Self>),
    IfThenElse(Ix<Self>, Ix<Self>, Ix<Self>),
    Name(Name),
    Block(Ix<Block>),
    Call(Callee, Vec<Ix<Self>>),
    Comptime(Ix<Self>),
    Struct(Ix<Struct>),
    Constructor(ConstructorType, Vec<StructField>),
    Field(Ix<Self>, Span<DefaultSymbol>),
    Error,
}

#[derive(Copy, Clone, Debug)]
pub enum Callee {
    Expr(Ix<Expr>),
    Builtin(Span<BuiltinFunction>),
}

/// None is anonymous constructor, Some is a given type
type ConstructorType = Option<Ix<Expr>>;

#[derive(Copy, Clone, Debug)]
pub enum Name {
    Let(Ix<Let>),
    Param(Ix<Param>),
    Function(Ix<Function>),
    Builtin(BuiltinType),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum BuiltinType {
    I32,
    Bool,
    Type,
}

#[derive(Copy, Clone, Debug)]
pub enum BuiltinFunction {
    InComptime,
    Trap,
    Clz,
    Ctz,
}

impl FromStr for BuiltinFunction {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "in_comptime" => Ok(BuiltinFunction::InComptime),
            "trap" => Ok(BuiltinFunction::Trap),
            "clz" => Ok(BuiltinFunction::Clz),
            "ctz" => Ok(BuiltinFunction::Ctz),
            _ => Err(()),
        }
    }
}

#[derive(Clone, Debug)]
pub struct Struct {
    pub fields: Vec<StructItem>,
}

#[derive(Clone, Debug)]
pub enum StructItem {
    Field(StructField),
}

#[derive(Clone, Debug)]
pub struct StructField {
    pub name: Span<DefaultSymbol>,
    pub value: Ix<Expr>,
}

impl BuiltinType {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::I32 => "i32",
            Self::Bool => "bool",
            Self::Type => "type",
        }
    }
}

impl FromStr for BuiltinType {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "i32" => Ok(Self::I32),
            "bool" => Ok(Self::Bool),
            "type" => Ok(Self::Type),
            _ => Err(()),
        }
    }
}

impl fmt::Display for BuiltinType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self.as_str(), f)
    }
}
