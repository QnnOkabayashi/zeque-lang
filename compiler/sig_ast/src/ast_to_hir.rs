//! The AST -> HIR transformation consists of:
//!   * name resolution
//!   * using an array and indices for the tree structure

use crate::hir::BuiltinType;
use crate::util::{Ix, Range, Scope, Span, StringInterner};
use crate::{ast, hir};
use smol_str::SmolStr;
use std::collections::HashMap;
use string_interner::DefaultSymbol;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum Error {
    #[error("name not found: `{0}`")]
    NameNotFound(SmolStr, Range),
    #[error("duplicate function names")]
    DuplicateFunctionNames(Vec<(DefaultSymbol, Vec<Range>)>),
    #[error("unknown builtin: `{0}`")]
    UnknownBuiltin(SmolStr, Range),
}

pub fn entry(functions: &[ast::Function]) -> Result<(Vec<hir::Function>, StringInterner), Error> {
    let mut interner = StringInterner::new();
    let mut hir_functions = Vec::with_capacity(functions.len());

    let mut env = Vec::with_capacity(functions.len());
    let mut program_scope = Scope::new(&mut env);

    let mut function_names: HashMap<DefaultSymbol, Vec<Range>> =
        HashMap::with_capacity(functions.len());
    for (index, function) in functions.iter().enumerate() {
        let Span(name, range) = &function.name;
        function_names
            .entry(interner.get_or_intern(name))
            .or_default()
            .push(*range);
        program_scope.push(hir::Name::Function(Ix::new(index)));
    }

    let duplicate_function_names: Vec<_> = function_names
        .into_iter()
        .filter(|(_, ranges)| ranges.len() > 1)
        .collect();
    if !duplicate_function_names.is_empty() {
        return Err(Error::DuplicateFunctionNames(duplicate_function_names));
    }

    for function in functions {
        let mut function_scope = program_scope.enter_scope();
        let mut lowerer = LoweringContext {
            context: hir::FunctionContext::default(),
            program: functions,
            builtin_string_symbols: BuiltinStringSymbols::new(&mut interner),
            interner: &mut interner,
        };

        for param in &function.params {
            lowerer.lower_param(param, &mut function_scope)?;
        }
        let Span(returns, _) = lowerer.lower_expr(&function.return_type, &mut function_scope)?;
        let Span(block_index, _) = lowerer.lower_block(&function.body, &mut function_scope)?;

        let symbol_index = lowerer.lower_str(&function.name);

        hir_functions.push(hir::Function {
            name: symbol_index,
            return_type: returns,
            body: block_index,
            context: lowerer.context,
        })
    }

    Ok((hir_functions, interner))
}

struct BuiltinStringSymbols {
    sym_i32: DefaultSymbol,
    sym_bool: DefaultSymbol,
    sym_type: DefaultSymbol,
}

impl BuiltinStringSymbols {
    fn new(interner: &mut StringInterner) -> Self {
        BuiltinStringSymbols {
            sym_i32: interner.get_or_intern(BuiltinType::I32.as_str()),
            sym_bool: interner.get_or_intern(BuiltinType::Bool.as_str()),
            sym_type: interner.get_or_intern(BuiltinType::Type.as_str()),
        }
    }
}

struct LoweringContext<'ast> {
    context: hir::FunctionContext,
    program: &'ast [ast::Function],
    builtin_string_symbols: BuiltinStringSymbols,
    interner: &'ast mut StringInterner,
}

impl<'ast> LoweringContext<'ast> {
    fn lower_str(&mut self, Span(name, range): &Span<SmolStr>) -> Span<DefaultSymbol> {
        let symbol = self.interner.get_or_intern(name);
        Span(symbol, *range)
    }

    fn lower_expr_raw(
        &mut self,
        expr: &ast::Expr,
        scope: &mut Scope<'_, hir::Name>,
    ) -> Result<Span<hir::Expr>, Error> {
        match expr {
            ast::Expr::Int(Span(int, range)) => Ok(Span(hir::Expr::Int(*int), *range)),
            ast::Expr::Bool(Span(boolean, range)) => Ok(Span(hir::Expr::Bool(*boolean), *range)),
            ast::Expr::BinOp(op, lhs, rhs) => {
                let Span(lhs, lhs_range) = self.lower_expr(lhs, scope)?;
                let Span(rhs, rhs_range) = self.lower_expr(rhs, scope)?;

                Ok(Span(
                    hir::Expr::BinOp(*op, lhs, rhs),
                    lhs_range.to(rhs_range),
                ))
            }
            ast::Expr::IfThenElse(ref cond, ref then, ref else_) => {
                let Span(cond, cond_range) = self.lower_expr(cond, scope)?;
                let Span(then, ..) = self.lower_expr(then, scope)?;
                let Span(else_, else_range) = self.lower_expr(else_, scope)?;

                Ok(Span(
                    hir::Expr::IfThenElse(cond, then, else_),
                    cond_range.to(else_range),
                ))
            }
            ast::Expr::Name(name) => {
                let Span(name, range) = self.lower_name(name.clone(), scope)?;

                Ok(Span(hir::Expr::Name(name), range))
            }
            ast::Expr::Call(callee, Span(arguments, args_range)) => {
                let Span(callee, callee_range) = match callee {
                    ast::Callee::Expr(callee) => {
                        let Span(callee_index, callee_range) = self.lower_expr(callee, scope)?;
                        Span(hir::Callee::Expr(callee_index), callee_range)
                    }
                    ast::Callee::Builtin(Span(builtin, range)) => {
                        let builtin = builtin
                            .as_str()
                            .parse()
                            .map_err(|()| Error::UnknownBuiltin(builtin.clone(), *range))?;

                        Span(hir::Callee::Builtin(Span(builtin, *range)), *range)
                    }
                };

                let argument_indices = arguments
                    .iter()
                    .map(|argument| self.lower_expr(argument, scope).map(Span::into_inner))
                    .collect::<Result<_, _>>()?;

                Ok(Span(
                    hir::Expr::Call(callee, argument_indices),
                    callee_range.to(*args_range),
                ))
            }
            ast::Expr::Block(block) => {
                let mut scope = scope.enter_scope();
                let Span(block_index, range) = self.lower_block(block, &mut scope)?;

                Ok(Span(hir::Expr::Block(block_index), range))
            }
            ast::Expr::Comptime(expr) => {
                let Span(expr_index, range) = self.lower_expr(expr, scope)?;

                Ok(Span(hir::Expr::Comptime(expr_index), range))
            }
            ast::Expr::Struct(struct_) => {
                let Span(struct_index, range) = self.lower_struct(struct_, scope)?;

                Ok(Span(hir::Expr::Struct(struct_index), range))
            }
            ast::Expr::Constructor(ctor, Span(fields, fields_range)) => {
                let Span(ctor, ctor_range) = self.lower_expr(ctor, scope)?;

                let fields = fields
                    .iter()
                    .map(|field| {
                        let name = self.lower_str(&field.name);
                        let Span(value, _) = self.lower_expr(&field.value, scope)?;

                        Ok(hir::StructField { name, value })
                    })
                    .collect::<Result<_, Error>>()?;

                Ok(Span(
                    hir::Expr::Constructor(Some(ctor), fields),
                    ctor_range.to(*fields_range),
                ))
            }
            ast::Expr::Field(value, field) => {
                let Span(value, value_range) = self.lower_expr(value, scope)?;
                let field = self.lower_str(field);

                Ok(Span(
                    hir::Expr::Field(value, field),
                    value_range.to(field.range()),
                ))
            }
            ast::Expr::Error(range) => Ok(Span(hir::Expr::Error, *range)),
        }
    }

    fn lower_expr(
        &mut self,
        expr: &ast::Expr,
        scope: &mut Scope<'_, hir::Name>,
    ) -> Result<Span<Ix<hir::Expr>>, Error> {
        let Span(expr, range) = self.lower_expr_raw(expr, scope)?;

        let index = self.context.exprs.push(Span(expr, range));
        Ok(Span(index, range))
    }

    fn lower_param(
        &mut self,
        Span(param, range): &Span<ast::Param>,
        scope: &mut Scope<'_, hir::Name>,
    ) -> Result<(), Error> {
        let Span(ty, _) = self.lower_expr(&param.ty.0, scope)?;

        let param = hir::Param {
            is_comptime: param.is_comptime.is_some(),
            name: self.lower_str(&param.name),
            ty,
        };

        let index = self.context.params.push(Span(param, *range));

        scope.push(hir::Name::Param(index));
        Ok(())
    }

    fn lower_block(
        &mut self,
        Span(block, range): &Span<ast::Block>,
        scope: &mut Scope<'_, hir::Name>,
    ) -> Result<Span<Ix<hir::Block>>, Error> {
        let mut stmts = Vec::with_capacity(block.stmts.len());
        for stmt in &block.stmts {
            stmts.push(self.lower_stmt(stmt, scope)?);
        }

        let Span(returns, _) = self.lower_expr(&block.returns, scope)?;
        let block = hir::Block { stmts, returns };

        Ok(Span(Ix::push(&mut self.context.blocks, block), *range))
    }

    fn lower_stmt(
        &mut self,
        Span(stmt, _range): &Span<ast::Stmt>,
        scope: &mut Scope<'_, hir::Name>,
    ) -> Result<hir::Stmt, Error> {
        match stmt {
            ast::Stmt::Let(let_) => {
                let Span(expr, _) = self.lower_expr(&let_.expr.0, scope)?;
                let ty = let_
                    .ty
                    .as_ref()
                    .map(|ty| self.lower_expr(&ty.0, scope).map(Span::into_inner))
                    .transpose()?;

                let let_ = hir::Let {
                    name: self.lower_str(&let_.name),
                    ty,
                    expr,
                };

                let index = Ix::push(&mut self.context.lets, let_);
                scope.push(hir::Name::Let(index));

                Ok(hir::Stmt::Let(index))
            }
        }
    }

    fn lower_name(
        &mut self,
        Span(name, range): Span<SmolStr>,
        scope: &mut Scope<'_, hir::Name>,
    ) -> Result<Span<hir::Name>, Error> {
        let symbol = self.interner.get_or_intern(&name);

        // Look for name in the current scope.
        // This includes other functions.
        for (_, name) in scope.iter() {
            if symbol == self.name_as_symbol(*name) {
                return Ok(Span(*name, range));
            }
        }

        if symbol == self.builtin_string_symbols.sym_i32 {
            Ok(Span(hir::Name::Builtin(hir::BuiltinType::I32), range))
        } else if symbol == self.builtin_string_symbols.sym_bool {
            Ok(Span(hir::Name::Builtin(hir::BuiltinType::Bool), range))
        } else if symbol == self.builtin_string_symbols.sym_type {
            Ok(Span(hir::Name::Builtin(hir::BuiltinType::Type), range))
        } else {
            Err(Error::NameNotFound(name, range))
        }
    }

    fn lower_struct(
        &mut self,
        Span(struct_, range): &Span<ast::Struct>,
        scope: &mut Scope<'_, hir::Name>,
    ) -> Result<Span<Ix<hir::Struct>>, Error> {
        let mut scope = scope.enter_scope();
        let fields = struct_
            .fields
            .iter()
            .map(|item| match item {
                ast::StructItem::Field(field) => {
                    let Span(value, _) = self.lower_expr(&field.value, &mut scope)?;

                    Ok(hir::StructItem::Field(hir::StructField {
                        name: self.lower_str(&field.name),
                        value,
                    }))
                }
            })
            .collect::<Result<Vec<_>, Error>>()?;

        Ok(Span(
            Ix::push(&mut self.context.structs, hir::Struct { fields }),
            *range,
        ))
    }

    fn name_as_symbol(&mut self, name: hir::Name) -> DefaultSymbol {
        match name {
            hir::Name::Let(index) => self.context.lets[index].name.0,
            hir::Name::Param(index) => self.context.params[index].name.0,
            hir::Name::Function(index) => self
                .interner
                .get_or_intern(&self.program[index.index].name.0),
            hir::Name::Builtin(builtin) => match builtin {
                BuiltinType::I32 => self.builtin_string_symbols.sym_i32,
                BuiltinType::Bool => self.builtin_string_symbols.sym_bool,
                BuiltinType::Type => self.builtin_string_symbols.sym_type,
            },
        }
    }
}
