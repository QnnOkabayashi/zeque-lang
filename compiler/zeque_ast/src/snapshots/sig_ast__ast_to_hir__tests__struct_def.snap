---
source: compiler/sig_ast/src/ast_to_hir.rs
expression: "parse(stringify!(a: struct { b: i32, },))"
---
File {
    struct_idx: 0,
    ctx: Ctx {
        params: [],
        exprs: [
            Name(
                BuiltinType(
                    I32,
                ),
            ),
            Struct(
                1,
            ),
        ],
        lets: [],
        structs: [
            Struct {
                fns: [],
                fields: [
                    FieldDecl {
                        name: "a",
                        ty: 1,
                    },
                ],
            },
            Struct {
                fns: [],
                fields: [
                    FieldDecl {
                        name: "b",
                        ty: 0,
                    },
                ],
            },
        ],
    },
}
